Mathlib/Algebra/Hom/Iterate.lean

```diff
@@ -0,0 +1,252 @@
+  f.toMultiplicative.iterate_map_pow n x m
+#align add_monoid_hom.iterate_map_smul AddMonoidHom.iterate_map_smul
+
+attribute [to_additive (reorder := 5)] MonoidHom.iterate_map_pow
+
+theorem iterate_map_zsmul (f : G →+ G) (n : ℕ) (m : ℤ) (x : G) : (f^[n]) (m • x) = m • (f^[n]) x :=
+  f.toMultiplicative.iterate_map_zpow n x m
+#align add_monoid_hom.iterate_map_zsmul AddMonoidHom.iterate_map_zsmul
+
+attribute [to_additive (reorder := 5)] MonoidHom.iterate_map_zpow
+
+end AddMonoidHom
+
+theorem AddMonoid.End.coe_pow {A} [AddMonoid A] (f : AddMonoid.End A) (n : ℕ) : ⇑(f ^ n) = f^[n] :=
+  hom_coe_pow _ rfl (fun _ _ => rfl) _ _
+#align add_monoid.End.coe_pow AddMonoid.End.coe_pow
+
+namespace RingHom
+
+section Semiring
+
+variable {R : Type _} [Semiring R] (f : R →+* R) (n : ℕ) (x y : R)
+
+theorem coe_pow (n : ℕ) : ⇑(f ^ n) = f^[n] :=
+  hom_coe_pow _ rfl (fun _ _ => rfl) f n
+#align ring_hom.coe_pow RingHom.coe_pow
+
+theorem iterate_map_one : (f^[n]) 1 = 1 :=
+  f.toMonoidHom.iterate_map_one n
+#align ring_hom.iterate_map_one RingHom.iterate_map_one
+
+theorem iterate_map_zero : (f^[n]) 0 = 0 :=
+  f.toAddMonoidHom.iterate_map_zero n
+#align ring_hom.iterate_map_zero RingHom.iterate_map_zero
+
+theorem iterate_map_add : (f^[n]) (x + y) = (f^[n]) x + (f^[n]) y :=
+  f.toAddMonoidHom.iterate_map_add n x y
+#align ring_hom.iterate_map_add RingHom.iterate_map_add
+
+theorem iterate_map_mul : (f^[n]) (x * y) = (f^[n]) x * (f^[n]) y :=
+  f.toMonoidHom.iterate_map_mul n x y
+#align ring_hom.iterate_map_mul RingHom.iterate_map_mul
+
+theorem iterate_map_pow (a) (n m : ℕ) : (f^[n]) (a ^ m) = (f^[n]) a ^ m :=
+  f.toMonoidHom.iterate_map_pow n a m
+#align ring_hom.iterate_map_pow RingHom.iterate_map_pow
+
+theorem iterate_map_smul (n m : ℕ) (x : R) : (f^[n]) (m • x) = m • (f^[n]) x :=
+  f.toAddMonoidHom.iterate_map_smul n m x
+#align ring_hom.iterate_map_smul RingHom.iterate_map_smul
+
+end Semiring
+
+variable {R : Type _} [Ring R] (f : R →+* R) (n : ℕ) (x y : R)
+
+theorem iterate_map_sub : (f^[n]) (x - y) = (f^[n]) x - (f^[n]) y :=
+  f.toAddMonoidHom.iterate_map_sub n x y
+#align ring_hom.iterate_map_sub RingHom.iterate_map_sub
+
+theorem iterate_map_neg : (f^[n]) (-x) = -(f^[n]) x :=
+  f.toAddMonoidHom.iterate_map_neg n x
+#align ring_hom.iterate_map_neg RingHom.iterate_map_neg
+
+theorem iterate_map_zsmul (n : ℕ) (m : ℤ) (x : R) : (f^[n]) (m • x) = m • (f^[n]) x :=
+  f.toAddMonoidHom.iterate_map_zsmul n m x
+#align ring_hom.iterate_map_zsmul RingHom.iterate_map_zsmul
+
+end RingHom
+
+theorem Equiv.Perm.coe_pow {α : Type _} (f : Equiv.Perm α) (n : ℕ) : ⇑(f ^ n) = f^[n] :=
+  hom_coe_pow _ rfl (fun _ _ => rfl) _ _
+#align equiv.perm.coe_pow Equiv.Perm.coe_pow
+
+--what should be the namespace for this section?
+section Monoid
+
+variable [Monoid G] (a : G) (n : ℕ)
+
+@[simp, to_additive]
+theorem smul_iterate [MulAction G H] : ((· • ·) a : H → H)^[n] = (· • ·) (a ^ n) :=
```

Comment:
`(· • ·) a` is not a faithful port of `(•) a` which means `Smul.smul a`

Reply:
Would you recommend `(· • ·) a` or `Smul.smul a` during port?

Reply:
I've changed them to `(· * a)`